session/txn.go

```diff
@@ -376,6 +376,13 @@ func (txn *LazyTxn) Rollback() error {
 	return txn.Transaction.Rollback()
 }
 
+// RollbackMemDBToCheckpoint overrides the Transaction interface.
+func (txn *LazyTxn) RollbackMemDBToCheckpoint(savepoint *tikv.MemDBCheckpoint) {
+	txn.flushStmtBuf()
+	txn.Transaction.RollbackMemDBToCheckpoint(savepoint)
+	txn.cleanup()
```

Comment:
The `cleanup` method would discard all the binlog mutations later. It's better to add comments to remind about it because it's not restored as other fields.

Reply:
Besides, there're still some transaction-related states or cache in `tikvTxn` and `KVTxn`, some of them could be affected by transaction executions. I'm not quite sure about the risks if they are not considered by the rollback process.

Reply:
`txn.cleanup` will only clean the binlog which is generated in current statement. for `rollback to savepoint` statement, the `txn.mutations` will always be nil. So `txn.cleanup` won't clear the old binlog.

But in another way, when binlog is enabled, execute `savepoint` statement will return an error.

Reply:
I have verified `tikvTxn` and `tikv.KVTxn`, there are not any other fields that need to be rollback.

Reply:
Got, maybe we need to add some defensive checks or tests in case some staging states are added in the future but they are not restored.